lib/task_manager.js

```diff
@@ -0,0 +1,80 @@
+class TaskManager {
+  constructor() {
+    this.queue;
+    this.buffer = [];
+    this.taskData = {};
+    this.taskIndex = 0;
+    this.currentFromBlock;
+  }
+
+  /**
+   * Method for initialization of the queue. It's done in such a way,
+   * because the p-queue package itself does not support CommonJS type importing.
+   * @param {number} maxConcurrentRequests Number of maximum concurrent tasks that should work at the same time
+   */
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+    this.queue.on('completed', ([interval, data]) => this.handleNewData(interval, data));
+  }
+
+  /**
+   * Method for creating a TaskManager instance.
+   * @param {number} maxConcurrentRequests Number of maximum concurrent tasks that should work at the same time
+   * @returns A TaskManager instance
+   */
+  static async create(maxConcurrentRequests) {
+    const tm = new TaskManager();
+    await tm.initQueue(maxConcurrentRequests);
+    return tm;
+  }
+
+  /**
+   * @returns A deep copy of the current TaskManager buffer
+   */
+  retrieveCompleted() {
+    const bufferCopy = [];
+    while (this.buffer.length > 0) bufferCopy.push(this.buffer.shift());
+    return bufferCopy;
+  }
+
+  /**
+   * Private method for pushing the sequential intervals that are ready.
+   * While the loop hits sequential intervals in the taskData property,
+   * the data should be pushed to the buffer property. When the while loop hits
+   * an undefined (yet) key of taskData, the function should stop.
+   */
+  #pushAllEligable() {
+    while (this.taskData[this.currentFromBlock]) {
+      for (const event of this.taskData[this.currentFromBlock].data) this.buffer.push(event);
+      const interval = this.taskData[this.currentFromBlock].toBlock - this.currentFromBlock + 1;
+      delete this.taskData[this.currentFromBlock];
+      this.currentFromBlock += interval;
+    }
+  }
+
+  /**
+   * On the completion of a task in the p-queue, the task
+   * should return an array in the form of [interval, events].
+   * These would be set up in the taskData object accordingly in the
+   * correct format, after which we use the #pushAllEligable private method
+   * to push ready-to-go sequential data into the buffer.
+   * @param {object} interval
+   * @param {Array} newTransformedData 
+   */
+  handleNewData(interval, newTransformedData) {
+    this.taskData[interval.fromBlock] = { toBlock: interval.toBlock, data: newTransformedData };
```

Comment:
I think at this point we need to have some blocking mechanism to block the producers. Currently we have no way to throttle the tasks coming from the PQueue and we would run into an OOM if the consumer of this data (the logic writing into Kafka) is slower. Right?